File: query.py
"""
RAG query utility for the InvestmentAdvisorRAG DynamoDB table.

Embeds a user question using Amazon Titan Text Embeddings V2, scans the DynamoDB table,
and returns the top-K most relevant chunks ranked by cosine similarity.

Usage:
    python rag/query.py "What is a Roth IRA?"
    python rag/query.py "How should a 30-year-old invest?" --top-k 5
    python rag/query.py "Tell me about bonds" --content-type FAQ
    python rag/query.py "Market outlook" --content-type MARKET_COMMENTARY --top-k 3
"""

import argparse
import json
import math
import os
from decimal import Decimal

import boto3
from boto3.dynamodb.conditions import Attr

TABLE_NAME = "InvestmentAdvisorRAG"
EMBED_MODEL_ID = "amazon.titan-embed-text-v2:0"


def get_region():
    return os.environ.get("AWS_REGION", os.environ.get("AWS_DEFAULT_REGION", "us-east-1"))


def get_bedrock_client():
    return boto3.client("bedrock-runtime", region_name=get_region())


def get_dynamo_table():
    dynamodb = boto3.resource("dynamodb", region_name=get_region())
    return dynamodb.Table(TABLE_NAME)


def embed_text(bedrock_client, text: str) -> list[float]:
    body = json.dumps({"inputText": text})
    response = bedrock_client.invoke_model(
        modelId=EMBED_MODEL_ID,
        body=body,
        contentType="application/json",
        accept="application/json",
    )
    result = json.loads(response["body"].read())
    return result["embedding"]


def cosine_similarity(a: list, b: list) -> float:
    """Compute cosine similarity between two vectors."""
    dot = sum(float(x) * float(y) for x, y in zip(a, b))
    norm_a = math.sqrt(sum(float(x) ** 2 for x in a))
    norm_b = math.sqrt(sum(float(x) ** 2 for x in b))
    if norm_a == 0 or norm_b == 0:
        return 0.0
    return dot / (norm_a * norm_b)


def scan_all_items(table, content_type: str | None = None) -> list[dict]:
    """Scan all items from DynamoDB, optionally filtered by content_type."""
    kwargs = {}
    if content_type:
        kwargs["FilterExpression"] = Attr("content_type").eq(content_type)

    items = []
    response = table.scan(**kwargs)
    items.extend(response.get("Items", []))

    while "LastEvaluatedKey" in response:
        kwargs["ExclusiveStartKey"] = response["LastEvaluatedKey"]
        response = table.scan(**kwargs)
        items.extend(response.get("Items", []))

    return items


def query(
    question: str,
    top_k: int = 5,
    content_type: str | None = None,
    show_scores: bool = True,
) -> list[dict]:
    """
    Embed a question, scan DynamoDB, and return top-K chunks by cosine similarity.

    Args:
        question: The user's query string.
        top_k: Number of results to return.
        content_type: Optional filter — PRODUCT | FAQ | MARKET_COMMENTARY | CLIENT_SCENARIO.
        show_scores: Whether to include similarity scores in output.

    Returns:
        List of dicts with keys: chunk_id, content_type, title, text, score, metadata.
    """
    bedrock = get_bedrock_client()
    table = get_dynamo_table()

    print(f"Embedding question...")
    query_embedding = embed_text(bedrock, question)

    print(f"Scanning table{f' (filter: {content_type})' if content_type else ''}...")
    items = scan_all_items(table, content_type)
    print(f"Retrieved {len(items)} items. Computing similarities...")

    scored = []
    for item in items:
        if "embedding" not in item:
            continue
        score = cosine_similarity(query_embedding, item["embedding"])
        scored.append({
            "chunk_id": item.get("chunk_id"),
            "content_type": item.get("content_type"),
            "title": item.get("title"),
            "text": item.get("text"),
            "metadata": item.get("metadata", {}),
            "score": score,
        })

    scored.sort(key=lambda x: x["score"], reverse=True)
    return scored[:top_k]


def format_results(results: list[dict], question: str) -> None:
    print(f"\n{'='*70}")
    print(f"Query: {question}")
    print(f"Top {len(results)} results:")
    print(f"{'='*70}\n")

    for i, r in enumerate(results, 1):
        print(f"[{i}] {r['title']}")
        print(f"    Type:  {r['content_type']}")
        print(f"    Score: {r['score']:.4f}")
        print(f"    Text:  {r['text'][:300]}{'...' if len(r['text']) > 300 else ''}")
        print()


def build_context(results: list[dict]) -> str:
    """Format retrieved chunks as context string for an LLM prompt."""
    context_parts = []
    for i, r in enumerate(results, 1):
        context_parts.append(
            f"[Source {i}: {r['title']} ({r['content_type']})]\n{r['text']}"
        )
    return "\n\n---\n\n".join(context_parts)


def query_and_answer(question: str, top_k: int = 5, content_type: str | None = None) -> str:
    """
    Full RAG pipeline: retrieve relevant chunks, then generate an answer using Claude.

    Returns the LLM's response string.
    """
    results = query(question, top_k=top_k, content_type=content_type)
    context = build_context(results)

    bedrock = boto3.client("bedrock-runtime", region_name=get_region())
    prompt = (
        f"You are a knowledgeable banking investment advisor assistant. "
        f"Use the following reference material to answer the client's question accurately and helpfully.\n\n"
        f"Reference Material:\n{context}\n\n"
        f"Client Question: {question}\n\n"
        f"Provide a clear, professional response. If the reference material doesn't fully address the question, "
        f"say so and provide general guidance based on sound financial principles."
    )

    body = {
        "anthropic_version": "bedrock-2023-05-31",
        "max_tokens": 1024,
        "messages": [{"role": "user", "content": prompt}],
    }

    response = bedrock.invoke_model(
        modelId="us.anthropic.claude-opus-4-6-v1",
        body=json.dumps(body),
        contentType="application/json",
        accept="application/json",
    )
    result = json.loads(response["body"].read())
    return result["content"][0]["text"]


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Query the investment advisor RAG system.")
    parser.add_argument("question", help="The question to ask.")
    parser.add_argument("--top-k", type=int, default=5, help="Number of results to return (default: 5).")
    parser.add_argument(
        "--content-type",
        choices=["PRODUCT", "FAQ", "MARKET_COMMENTARY", "CLIENT_SCENARIO"],
        default=None,
        help="Filter results by content type.",
    )
    parser.add_argument(
        "--answer",
        action="store_true",
        help="Generate a full LLM answer using retrieved chunks as context.",
    )
    args = parser.parse_args()

    if args.answer:
        print(f"\nGenerating RAG answer for: {args.question}\n")
        answer = query_and_answer(args.question, top_k=args.top_k, content_type=args.content_type)
        print(f"{'='*70}")
        print("Answer:")
        print(f"{'='*70}")
        print(answer)
    else:
        results = query(args.question, top_k=args.top_k, content_type=args.content_type)
        format_results(results, args.question)

File: query.js
/**
 * RAG query utility for the InvestmentAdvisorRAG DynamoDB table.
 *
 * Embeds a user question using Amazon Titan Text Embeddings V2, scans the DynamoDB table,
 * and returns the top-K most relevant chunks ranked by cosine similarity.
 *
 * Usage:
 *   node rag/query.js "What is a Roth IRA?"
 *   node rag/query.js "How should a 30-year-old invest?" --top-k 5
 *   node rag/query.js "Tell me about bonds" --content-type FAQ
 *   node rag/query.js "Market outlook" --content-type MARKET_COMMENTARY --top-k 3
 *   node rag/query.js "What should I do with an inheritance?" --answer
 */

import { BedrockRuntimeClient, InvokeModelCommand } from "@aws-sdk/client-bedrock-runtime";
import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
import { DynamoDBDocumentClient, ScanCommand } from "@aws-sdk/lib-dynamodb";
import { readFileSync } from "fs";
import { parseArgs } from "util";

// Load .env file from parent directory if present
try {
  const envPath = new URL("../.env", import.meta.url).pathname;
  const envContent = readFileSync(envPath, "utf8");
  for (const line of envContent.split("\n")) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith("#")) continue;
    const eqIdx = trimmed.indexOf("=");
    if (eqIdx === -1) continue;
    const key = trimmed.slice(0, eqIdx).trim();
    const value = trimmed.slice(eqIdx + 1).trim();
    if (key && !(key in process.env)) {
      process.env[key] = value;
    }
  }
} catch {
  // No .env file — rely on environment variables already set
}

const TABLE_NAME = "InvestmentAdvisorRAG";
const EMBED_MODEL_ID = "amazon.titan-embed-text-v2:0";
const CLAUDE_MODEL_ID = "us.anthropic.claude-opus-4-6-v1";

function getRegion() {
  return process.env.AWS_REGION ?? process.env.AWS_DEFAULT_REGION ?? "us-east-1";
}

function getBedrockClient() {
  return new BedrockRuntimeClient({ region: getRegion() });
}

function getDynamoClient() {
  const dynamo = new DynamoDBClient({ region: getRegion() });
  return DynamoDBDocumentClient.from(dynamo, {
    marshallOptions: { convertClassInstanceToMap: true },
  });
}

/**
 * Embed a text string using Amazon Titan Text Embeddings V2.
 * @param {BedrockRuntimeClient} client
 * @param {string} text
 * @returns {Promise<number[]>}
 */
async function embedText(client, text) {
  const command = new InvokeModelCommand({
    modelId: EMBED_MODEL_ID,
    body: JSON.stringify({ inputText: text }),
    contentType: "application/json",
    accept: "application/json",
  });
  const response = await client.send(command);
  const result = JSON.parse(Buffer.from(response.body).toString("utf8"));
  return result.embedding;
}

/**
 * Compute cosine similarity between two numeric vectors.
 * @param {number[]} a
 * @param {number[]} b
 * @returns {number}
 */
function cosineSimilarity(a, b) {
  let dot = 0;
  let normA = 0;
  let normB = 0;
  for (let i = 0; i < a.length; i++) {
    const ai = Number(a[i]);
    const bi = Number(b[i]);
    dot += ai * bi;
    normA += ai * ai;
    normB += bi * bi;
  }
  if (normA === 0 || normB === 0) return 0;
  return dot / (Math.sqrt(normA) * Math.sqrt(normB));
}

/**
 * Scan all items from DynamoDB, optionally filtered by content_type.
 * Handles pagination automatically.
 * @param {DynamoDBDocumentClient} client
 * @param {string|null} contentType
 * @returns {Promise<object[]>}
 */
async function scanAllItems(client, contentType = null) {
  const items = [];
  let lastKey = undefined;

  do {
    const params = {
      TableName: TABLE_NAME,
      ...(contentType && {
        FilterExpression: "content_type = :ct",
        ExpressionAttributeValues: { ":ct": contentType },
      }),
      ...(lastKey && { ExclusiveStartKey: lastKey }),
    };

    const response = await client.send(new ScanCommand(params));
    items.push(...(response.Items ?? []));
    lastKey = response.LastEvaluatedKey;
  } while (lastKey);

  return items;
}

/**
 * Embed a question, scan DynamoDB, and return top-K chunks by cosine similarity.
 * @param {string} question
 * @param {object} options
 * @param {number} [options.topK=5]
 * @param {string|null} [options.contentType=null]
 * @returns {Promise<object[]>}
 */
async function query(question, { topK = 5, contentType = null } = {}) {
  const bedrock = getBedrockClient();
  const dynamo = getDynamoClient();

  console.log("Embedding question...");
  const queryEmbedding = await embedText(bedrock, question);

  const filterLabel = contentType ? ` (filter: ${contentType})` : "";
  console.log(`Scanning table${filterLabel}...`);
  const items = await scanAllItems(dynamo, contentType);
  console.log(`Retrieved ${items.length} items. Computing similarities...`);

  const scored = items
    .filter((item) => item.embedding)
    .map((item) => ({
      chunk_id: item.chunk_id,
      content_type: item.content_type,
      title: item.title,
      text: item.text,
      metadata: item.metadata ?? {},
      score: cosineSimilarity(queryEmbedding, item.embedding),
    }));

  scored.sort((a, b) => b.score - a.score);
  return scored.slice(0, topK);
}

/**
 * Format and print results to stdout.
 * @param {object[]} results
 * @param {string} question
 */
function formatResults(results, question) {
  console.log("\n" + "=".repeat(70));
  console.log(`Query: ${question}`);
  console.log(`Top ${results.length} results:`);
  console.log("=".repeat(70) + "\n");

  for (let i = 0; i < results.length; i++) {
    const r = results[i];
    const preview = r.text.length > 300 ? r.text.slice(0, 300) + "..." : r.text;
    console.log(`[${i + 1}] ${r.title}`);
    console.log(`    Type:  ${r.content_type}`);
    console.log(`    Score: ${r.score.toFixed(4)}`);
    console.log(`    Text:  ${preview}`);
    console.log();
  }
}

/**
 * Format retrieved chunks as a context string for an LLM prompt.
 * @param {object[]} results
 * @returns {string}
 */
function buildContext(results) {
  return results
    .map((r, i) => `[Source ${i + 1}: ${r.title} (${r.content_type})]\n${r.text}`)
    .join("\n\n---\n\n");
}

/**
 * Full RAG pipeline: retrieve relevant chunks, then generate an answer using Claude.
 * @param {string} question
 * @param {object} options
 * @param {number} [options.topK=5]
 * @param {string|null} [options.contentType=null]
 * @returns {Promise<string>}
 */
async function queryAndAnswer(question, { topK = 5, contentType = null } = {}) {
  const results = await query(question, { topK, contentType });
  const context = buildContext(results);

  const bedrock = getBedrockClient();
  const prompt =
    `You are a knowledgeable banking investment advisor assistant. ` +
    `Use the following reference material to answer the client's question accurately and helpfully.\n\n` +
    `Reference Material:\n${context}\n\n` +
    `Client Question: ${question}\n\n` +
    `Provide a clear, professional response. If the reference material doesn't fully address the question, ` +
    `say so and provide general guidance based on sound financial principles.`;

  const body = {
    anthropic_version: "bedrock-2023-05-31",
    max_tokens: 1024,
    messages: [{ role: "user", content: prompt }],
  };

  const command = new InvokeModelCommand({
    modelId: CLAUDE_MODEL_ID,
    body: JSON.stringify(body),
    contentType: "application/json",
    accept: "application/json",
  });

  const response = await bedrock.send(command);
  const result = JSON.parse(Buffer.from(response.body).toString("utf8"));
  return result.content[0].text;
}

// ---------------------------------------------------------------------------
// CLI entry point
// ---------------------------------------------------------------------------

const VALID_CONTENT_TYPES = ["PRODUCT", "FAQ", "MARKET_COMMENTARY", "CLIENT_SCENARIO"];

const { values, positionals } = parseArgs({
  args: process.argv.slice(2),
  options: {
    "top-k": { type: "string", default: "5" },
    "content-type": { type: "string" },
    answer: { type: "boolean", default: false },
  },
  allowPositionals: true,
});

const question = positionals[0];
if (!question) {
  console.error('Usage: node rag/query.js "<question>" [--top-k N] [--content-type TYPE] [--answer]');
  console.error(`  --content-type: one of ${VALID_CONTENT_TYPES.join(", ")}`);
  process.exit(1);
}

const topK = parseInt(values["top-k"], 10);
const contentType = values["content-type"] ?? null;

if (contentType && !VALID_CONTENT_TYPES.includes(contentType)) {
  console.error(`Invalid --content-type "${contentType}". Must be one of: ${VALID_CONTENT_TYPES.join(", ")}`);
  process.exit(1);
}

if (values.answer) {
  console.log(`\nGenerating RAG answer for: ${question}\n`);
  const answer = await queryAndAnswer(question, { topK, contentType });
  console.log("=".repeat(70));
  console.log("Answer:");
  console.log("=".repeat(70));
  console.log(answer);
} else {
  const results = await query(question, { topK, contentType });
  formatResults(results, question);
}
